superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+
+    // restore all ips from char* to vector<string>
+    char *str_idx = all_str;
+    for (int rank_idx = 0; rank_idx < world_size; rank_idx++) {
+        ips.push_back(str_idx);
+        str_idx = str_idx + max_len + 1;
+    }
+}
+
+// Parse raw output of ib command
+float process_raw_output(string output) {
+    float res = -1.0;
+    try {
+        vector<string> lines;
+        split_string(output, lines, "\n");
+        regex re("\\d+\\s+\\d+\\s+\\d+\\.\\d+\\s+(\\d+\\.\\d+)\\s+\\d+\\.\\d+");
+        for (string line : lines) {
+            smatch m;
+            if (regex_search(line, m, re))
+                res = std::max(res, stof(m.str(1)));
+        }
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse raw_output: " << output << std::endl;
+    }
+
+    return res;
+}
+
+// Run ib command on server/client with server ip
+float run_cmd(string cmd_prefix, int port, bool server, string ip) {
+    string is_server = server ? "server" : "client";
+    // client sleep 1s in case that client starts before server
+    if (!server) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    }
+
+    string command, output;
+    try {
+        // exec command in termimal
+        command = cmd_prefix + " -p " + to_string(port);
+        command = server ? command : command + " " + ip;
+        output = exec(command.c_str());
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to exec command: " << command << std::endl;
+    }
+
+    // parse raw output and get result
+    float result = process_raw_output(output);
+
+    return result;
+}
+
+// The ranks in vector of (server, client) run commands parallel
+vector<float> run_cmd_parallel(string cmd_prefix, const vector<std::pair<int, int>> &run_pairs_in_parallel,
+                               const vector<int> &ports, const vector<string> &ips) {
+    int size = run_pairs_in_parallel.size();
+    // invoke function to run cmd in multi threads mode for each rank in the pairs
+    unordered_map<int, std::future<float>> threads;
+    for (int index = 0; index < size; index++) {
+        int server_index = run_pairs_in_parallel[index].first;
+        int client_index = run_pairs_in_parallel[index].second;
+        if (server_index == world_rank) {
+            threads[2 * index] = (std::async(run_cmd, cmd_prefix, ports[index], true, ips[server_index]));
+        }
+        if (client_index == world_rank) {
+            threads[2 * index + 1] = (std::async(run_cmd, cmd_prefix, ports[index], false, ips[server_index]));
+        }
+    }
+    // send the result of client to rank0
+    for (auto &thread : threads) {
+        float client_result = thread.second.get();
+        int index = thread.first;
+        if (index % 2 == 1) {
+            MPI_Send(&client_result, 1, MPI_FLOAT, 0, index, MPI_COMM_WORLD);
+        }
+    }
+    // rank 0 recv results
+    vector<float> results;
+    if (world_rank == 0) {
```

Comment:
nit: it will be better to push this up (even before launching the tests) to avoid unexpected MPI messages.